### PR TITLE
Move evaluation results to separate classes

### LINKED ISSUE #1280: Evaluation result classes shouldn't be nested under the Evaluator classes
When using our evaluation API:

```C#
BinaryClassificationContext binClassificationCtx = new BinaryClassificationContext(env);
BinaryClassifierEvaluator.CalibratedResult metrics = binClassificationCtx.Evaluate(predictions, "Label");
```

It is quite odd for the `metrics` object to be a nested class in the `BinaryClassifierEvaluator`. I didn't even use the `BinaryClassifierEvaluator` class at all, why am I dealing with a nested class inside of it?

I think these result classes should be moved to no longer be nested classes. Instead we should just have a `BinaryClassifierEvaluationResult` class.

/cc @Zruty0 @TomFinley @KrzysztofCwalina @terrajobst

### PR DESCRIPTION
**Pull request by robosek**
Fixes #1280 

NOTE:
- [x]  Moved results from (Ranking, Multiclass, Regression, Clustering, Binary) evaluators to separte classes under `Evaluators/Metrics` directory.
- [x] Adjusted types in dependent projects classes.
- [x] Build `Release` and `Debug` successful on local :+1:

### CODE REVIEW

**Comment by robosek** [comment]
@Zruty0 You mean I should change only the namespaces to `Microsoft.ML.Data` of new classes right?

**Comment by Zruty0** [comment]
@robosek yes, that's what I meant

File: src/Microsoft.ML.Data/Evaluators/Metrics/ClusteringMetrics.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        /// <summary>
+        /// Normalized Mutual Information
+        /// NMI is a measure of the mutual dependence of the variables.
+        /// <a href="http://en.wikipedia.org/wiki/Mutual_information#Normalized_variants">Normalized variants</a> work on data that already has cluster labels.
+        /// Its value ranged from 0 to 1, where higher numbers are better.
+        /// </summary>
+        public double Nmi { get; }
+
+        /// <summary>
+        /// Average Score. For the K-Means algorithm, the 'score' is the distance from the centroid to the example.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/Metrics/ClusteringMetrics.cs]: >' [](start = 58, length = 1)

The apostrophes were already there, but we might need to XML encode them, here and below.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs [hunk 2]
```diff
@@ -162,7 +162,8 @@ private BinaryClassificationMetrics EvaluateBinary(IHostEnvironment env, IDataVi
             var evaluator = new BinaryClassifierMamlEvaluator(env, new BinaryClassifierMamlEvaluator.Arguments());
             var metricsDic = evaluator.Evaluate(dataEval);
 
-            return BinaryClassificationMetrics.FromMetrics(env, metricsDic["OverallMetrics"], metricsDic["ConfusionMatrix"])[0];
+            return Microsoft.ML.Legacy.Models.BinaryClassificationMetrics
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs]: >Microsoft.ML.Legacy.Models [](start = 19, length = 26)

does this need to be Microsoft.ML.Data?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs [hunk 3]
```diff
@@ -151,7 +151,7 @@ public void TrainAndPredictSentimentModelWithDirectionInstantiationTestWithWordE
             Assert.Equal(1.0, (double)summary[0].Value, 1);
         }
 
-        private BinaryClassificationMetrics EvaluateBinary(IHostEnvironment env, IDataView scoredData)
+        private Microsoft.ML.Legacy.Models.BinaryClassificationMetrics EvaluateBinary(IHostEnvironment env, IDataView scoredData)
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs]: >Microsoft.ML.Legacy.Models [](start = 16, length = 26)

Microsoft.ML.Data?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs (hunk 2 above)
Comment (robosek) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs]: Actually before my change it was `BinaryClassificationMetrics` from `Microsoft.ML.Legacy.Models` but I had to add full name of class to distinguish between new `BinaryClassificationMetrics` from `Micorosft.ML.Data` and the legacy one.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs (hunk 3 above)
Comment (robosek) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/SentimentPredictionTests.cs]: The same as above

File: src/Microsoft.ML.Data/Evaluators/Metrics/ClusteringMetrics.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        /// Average Score. For the K-Means algorithm, the &apos;score&apos; is the distance from the centroid to the example.
+        /// The average score is, therefore, a measure of proximity of the examples to cluster centroids.
+        /// In other words, it&apos;s the &apos;cluster tightness&apos; measure.
+        /// Note however, that this metric will only decrease if the number of clusters is increased,
+        /// and in the extreme case (where each distinct example is its own cluster) it will be equal to zero.
+        /// </summary>
+        public double AvgMinScore { get; }
+
+        /// <summary>
+        /// <a href="https://en.wikipedia.org/wiki/Davies�Bouldin_index">Davies-Bouldin Index</a>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/ClusteringMetrics.cs]: >� [](start = 57, length = 1)

something went wrong with this hypen. I can't open link.
`[https://en.wikipedia.org/wiki/Davies–Bouldin_index`](https://en.wikipedia.org/wiki/Davies–Bouldin_index`) works fine

File: src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Data
+{
+    public sealed class RankerMetrics
+    {
+        /// <summary>
+        /// Normalized Discounted Cumulative Gain
+        /// <a href="https://github.com/dotnet/machinelearning/tree/master/docs/images/ndcg.png"></a>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs]: >ndcg [](start = 87, length = 4)

need to be `NDCG.png`

File: src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        public double[] Ndcg { get; }
+
+        /// <summary>
+        /// <a href="https://en.wikipedia.org/wiki/Discounted_cumulative_gain">Discounted Cumulative gain</a>
+        /// is the sum of the gains, for all the instances i, normalized by the natural logarithm of the instance + 1.
+        /// Note that unline the Wikipedia article, ML.Net uses the natural logarithm.
+        /// <a href="https://github.com/dotnet/machinelearning/tree/master/docs/images/dcg.png"></a>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs]: >dcg.png [](start = 87, length = 7)

need to be `DCG.png`

File: src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Data
+{
+    public sealed class RankerMetrics
+    {
+        /// <summary>
+        /// Normalized Discounted Cumulative Gain
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs]: >Normalized Discounted Cumulative Gain [](start = 12, length = 37)

Array of normalized discounted cumulative gains where i-th element represent NDCG@i (you can also phrase it as NDCG at i).

File: src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        /// <a href="https://github.com/dotnet/machinelearning/tree/master/docs/images/ndcg.png"></a>
+        /// </summary>
+        public double[] Ndcg { get; }
+
+        /// <summary>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs]: Array of discounted cumulative gains where i-th element represent DCG@i (you can also phrase it as DCG at i).

File: src/Microsoft.ML.Data/Evaluators/Metrics/RegressionMetrics.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        /// </remarks>
+        public double L2 { get; }
+
+        /// <summary>
+        /// Gets the root mean square loss (or RMS) which is the square root of the L2 loss.
+        /// </summary>
+        public double Rms { get; }
+
+        /// <summary>
+        /// Gets the user defined loss function.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RegressionMetrics.cs]: >Gets the user [](start = 12, length = 13)

Gets the result of user defined loss function.

File: src/Microsoft.ML.Data/Training/TrainingStaticExtensions.cs [hunk 10]
```diff
@@ -11,6 +11,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using Microsoft.ML.Data;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Training/TrainingStaticExtensions.cs]: would be nice to sort them.

**Comment by Ivanidzo4ka** [comment]
@robosek  thank you for your contribution!
I left few comments, and you need to resolve some merge conflicts (it's end of cycle time, so people tend to push lot of PRs)

**Comment by robosek** [comment]
@Ivanidzo4ka thank you for your review. Descriptions are updated and conflicts are resolved (at least for a while :smiley:)

**Comment by Ivanidzo4ka** [comment]
@robosek can you do following:
open command line and go to your machinelearning repository.
`git submodule update --remote --merge`
you should get following message in the end:

Submodule path 'src/Native/MatrixFactorizationNative/libmf': merged in '53a91e7e8c88463e97acfbbafb7134a6030860b3'

stage it and merge into your PR.

We have submodule which we recently updated in master, but in your branch it's still looking on old one, and for some reason git not smart enough to merge it properly.
Command above will point on latest master branch in libmf.

**Comment by robosek** [comment]
@Ivanidzo4ka I did before but this time I forgot to merge submodules. Sorry! Now it's merged as well :+1:

File: src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Data
+{
+    public sealed class RankerMetrics
+    {
+        /// <summary>
+        /// Array of normalized discounted cumulative gains where i-th element represent NDCG@i (you can also phrase it as NDCG at i).
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Evaluators/Metrics/RankerMetrics.cs]: >(you can also phrase it as NDCG at i) [](start = 96, length = 37)

You don't have to write this. I just try to give you option how you can phrase it :)
it's either NDCG@i or NGDCG at i, both a valid names.

**Comment by robosek** [comment]
@Ivanidzo4ka of course. Done :stuck_out_tongue:

**Pull request merged**